cli/popper/parser.py

```diff
@@ -198,53 +203,110 @@ def normalize(self):
+        if not isinstance(self._workflow['on'], basestring):
             log.fail('[on] attribute must be a string')
-        for _, a_block in self.workflow['action'].items():
+        for a_name, a_block in self._workflow['action'].items():
+
+            a_block['name'] = a_name
+
             if not isinstance(a_block['uses'], basestring):
                 log.fail('[uses] attribute must be a string')
             if a_block.get('needs', None):
                 if isinstance(a_block['needs'], basestring):
                     a_block['needs'] = [a_block['needs']]
-                elif not self.is_list_of_strings(a_block['needs']):
+                elif not self._is_list_of_strings(a_block['needs']):
                     log.fail(
                         '[needs] attribute must be a list of strings '
                         'or a string')
             if a_block.get('runs', None):
                 if isinstance(a_block['runs'], basestring):
                     a_block['runs'] = [a_block['runs']]
-                elif not self.is_list_of_strings(a_block['runs']):
+                elif not self._is_list_of_strings(a_block['runs']):
                     log.fail(
                         '[runs] attribute must be a list of strings '
                         'or a string')
             if a_block.get('args', None):
                 if isinstance(a_block['args'], basestring):
                     a_block['args'] = a_block['args'].split()
-                elif not self.is_list_of_strings(a_block['args']):
+                elif not self._is_list_of_strings(a_block['args']):
                     log.fail(
                         '[args] attribute must be a list of strings '
                         'or a string')
             if a_block.get('env', None):
                 if not isinstance(a_block['env'], dict):
                     log.fail('[env] attribute must be a dict')
             if a_block.get('secrets', None):
-                if not self.is_list_of_strings(a_block['secrets']):
+                if not self._is_list_of_strings(a_block['secrets']):
                     log.fail('[secrets] attribute must be a list of strings')
 
-    def check_duplicate_actions(self):
+    def _check_duplicate_actions(self):
         """Checks whether duplicate action blocks are
         present or not."""
         parsed_acount = 0
-        if self.workflow.get('action', None):
-            parsed_acount = len(list(self.workflow['action'].items()))
+        if self._workflow.get('action', None):
+            parsed_acount = len(list(self._workflow['action'].items()))
         acount = 0
-        for line in self.workflow_content:
+        for line in self._workflow_content:
             line = line.strip()
             if line.startswith('action '):
                 acount += 1
         if parsed_acount != acount:
             log.fail('Duplicate action identifiers found.')
+
+
+def validate_workflow(workflow):
```

Comment:
this can also be a member function, or is there a particular reason this is not a member function? Also, `validate_workflow` might be a bit ambiguous. We could rename this to `find_unreachable_actions` or `check_for_unreachable_actions`